search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+                              const gchar     *name,
+                              gpointer         user_data)
+{
+    g_debug ("Search provider name lost: %s\n", name);
+}
+
+static void
+search_provider_bus_acquired_cb (GDBusConnection *connection,
+                                 const gchar *name,
+                                 gpointer user_data)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    self->object_manager = g_dbus_object_manager_server_new ("/org/gnome/Yelp/SearchProvider");
+    self->skeleton = yelp_shell_search_provider2_skeleton_new ();
+
+    g_signal_connect (self->skeleton, "handle-get-initial-result-set",
+                    G_CALLBACK (handle_get_initial_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-subsearch-result-set",
+                    G_CALLBACK (handle_get_subsearch_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                    G_CALLBACK (handle_get_result_metas), self);
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                    G_CALLBACK (handle_activate_result), self);
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                    G_CALLBACK (handle_launch_search), self);
+
+    g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (self->skeleton),
+                                      connection,
+                                      "/org/gnome/Yelp/SearchProvider",
+                                      NULL);
+    g_dbus_object_manager_server_set_connection (self->object_manager, connection);
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+    if (self->name_owner_id != 0) {
+        g_bus_unown_name (self->name_owner_id);
+        self->name_owner_id = 0;
+    }
+
+    if (self->skeleton != NULL) {
+        g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON (self->skeleton));
+        g_clear_object (&self->skeleton);
+    }
+
+    g_clear_object (&self->object_manager);
+    g_clear_object (&self->doc);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                       YelpDocumentSignal    signal,
+                       YelpSearchProviderApp *app,
+                       GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    for (iter = page_ids; *iter; iter++){
+        gchar *page_id = *iter;
+        PageData *data = g_new0 (PageData, 1);
+        gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+        GIcon *icon = g_icon_new_for_string (icon_string, &error);
```

Comment:
I would just pass NULL here I think, because you do not do anything with the error. But if you want to print an error (with g_warning or g_debug) then you need to pass your own instance of a GError here, because the `error` variable comes from libyelp and is meant to be used for reporting an error from libyelp.